core/commands/object.go

```diff
@@ -638,31 +647,33 @@ func addLink(ctx context.Context, ds dag.DAGService, root *dag.Node, childname s
 	return root, nil
 }
 
-func insertNodeAtPath(ctx context.Context, ds dag.DAGService, root *dag.Node, path []string, toinsert key.Key) (*dag.Node, error) {
+func insertNodeAtPath(ctx context.Context, ds dag.DAGService, root *dag.Node, path []string, toinsert key.Key, create bool) (*dag.Node, error) {
 	if len(path) == 1 {
 		return addLink(ctx, ds, root, path[0], toinsert)
 	}
 
+	var nd *dag.Node
 	child, err := root.GetNodeLink(path[0])
 	if err != nil {
-		return nil, err
-	}
-
-	nd, err := child.GetNode(ctx, ds)
-	if err != nil {
-		return nil, err
-	}
-
-	ndprime, err := insertNodeAtPath(ctx, ds, nd, path[1:], toinsert)
-	if err != nil {
-		return nil, err
+		// if 'create' is true, we create directories on the way down as needed
+		if err == dag.ErrNotFound && create {
+			nd = &dag.Node{Data: ft.FolderPBData()}
+		} else {
+			return nil, err
+		}
+	} else {
+		nd, err = child.GetNode(ctx, ds)
+		if err != nil {
+			return nil, err
+		}
 	}
 
-	err = root.RemoveNodeLink(path[0])
+	ndprime, err := insertNodeAtPath(ctx, ds, nd, path[1:], toinsert, create)
```

Comment:
looks like this recursive construction works. but would love to see it tested more thoroughly.